arroyo-df/src/physical.rs

```diff
@@ -156,7 +154,119 @@ pub fn window_scalar_function() -> ScalarUDF {
     )
 }
 
-impl FunctionRegistry for EmptyRegistry {
+#[derive(WrapperApi)]
+struct UdfDylibInterface {
+    run: unsafe extern "C" fn(
+        args: Vec<(FFI_ArrowArray, FFI_ArrowSchema)>,
+    ) -> (FFI_ArrowArray, FFI_ArrowSchema),
+}
+
+pub struct UdfDylib {
+    name: String,
+    signature: Signature,
+    return_type: DataType,
+    udf: Container<UdfDylibInterface>,
+}
+
+impl Debug for UdfDylib {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("UdfDylib").finish()
+    }
+}
+
+impl UdfDylib {
+    pub async fn new(name: &str, config: &DylibUdfConfig) -> Self {
+        let signature = Signature::exact(
+            config
+                .arg_types
+                .iter()
+                .map(|arg| arrow_type_to_data_type(arg))
+                .collect(),
+            Volatility::Volatile,
+        );
+
+        let artifact_url =
+            std::env::var(ARTIFACT_URL_ENV).expect(&format!("{} must be set", ARTIFACT_URL_ENV));
+
+        let storage = StorageProvider::for_url(&artifact_url)
+            .await
+            .expect("unable to construct storage provider");
+
+        let udf = storage
+            .get(&config.dylib_path)
+            .await
+            .expect("UDF dylib not found.");
+
+        // write the dylib to a local file
+        let local_udfs_dir = "/tmp/arroyo/local_udfs";
+        std::fs::create_dir_all(local_udfs_dir).expect("unable to create local udfs dir");
+        let local_dylib_path = Path::new(local_udfs_dir).join(dylib_name(name));
+
+        std::fs::write(&local_dylib_path, udf).expect("unable to write dylib to file");
+
+        Self {
+            name: name.to_string(),
+            signature,
+            udf: unsafe { Container::load(local_dylib_path).unwrap() },
+            return_type: arrow_type_to_data_type(&config.return_type),
+        }
+    }
+}
+
+impl ScalarUDFImpl for UdfDylib {
+    fn as_any(&self) -> &dyn Any {
+        self
+    }
+
+    fn name(&self) -> &str {
+        &self.name
+    }
+
+    fn signature(&self) -> &Signature {
+        &self.signature
+    }
+
+    fn return_type(&self, _arg_types: &[DataType]) -> DFResult<DataType> {
+        Ok(self.return_type.clone())
+    }
+
+    fn invoke(&self, args: &[ColumnarValue]) -> DFResult<ColumnarValue> {
+        let args = args
+            .iter()
+            .map(|arg| match arg {
+                ColumnarValue::Array(array) => to_ffi(&(array.to_data())).unwrap(),
+                ColumnarValue::Scalar(_) => {
+                    unimplemented!("Scalar args not supported")
```

Comment:
We probably will need to support this or prevent it from planning (for example "select udf(5) from impulse;" will trigger this case)

Reply:
I've implemented this with a function `scalar_value_to_array`.